Issue #805

hyc opened the pull request.

moneromooo-monero commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -1824,18 +1824,19 @@ bool simple_wallet::refresh_main(uint64_t start_height, bool reset)
-    catch(const boost::bad_lexical_cast &)
+    catch(const boost::bad_lexical_cast &ignored)
     {
-        start_height = 0;
```
May want to add a m_wallet->set_refresh_from_block_height(0) here

hyc replied:
good point. done.

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -772,6 +790,60 @@ void wallet2::check_pending_txes()
+  size_t current_index = m_blockchain.size();
+  while(current_index < m_refresh_from_block_height)
+  {
+    pull_hashes(start_height, blocks_start_height, short_chain_history, hashes);
```
Seems start_height never gets increased by hashes.size() for a second loop

hyc replied:
That's intentional. Perhaps I should just get rid of the start_height parameter. (pull_hashes was just a copy/paste of pull_blocks.) If you actually increment the start_height parameter, everything breaks, same as running refresh with a non-zero start_height.

hyc replied:
Then again, I don't see why it should be broken like this. I think this is actually the same problem with the original refresh <height> - pull_next_blocks() never increments start_height either, so it pulls the same chunk of blocks over and over.

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -578,6 +578,24 @@ void wallet2::pull_blocks(uint64_t start_height, uint64_t &blocks_start_height,
+  m_daemon_rpc_mutex.unlock();
+  THROW_WALLET_EXCEPTION_IF(!r, error::no_connection_to_daemon, "gethashes.bin");
+  THROW_WALLET_EXCEPTION_IF(res.status == CORE_RPC_STATUS_BUSY, error::daemon_busy, "gethashes.bin");
+  THROW_WALLET_EXCEPTION_IF(res.status != CORE_RPC_STATUS_OK, error::get_blocks_error, res.status);
```
Probably best to make a separate exception, those are just a specialization of a base, so just a couple lines to add to src/wallet/wallet_errors.h

moneromooo-monero commented:
looks ok modulo the comments

hyc commented:
After I figured out what was wrong with the original "refresh <height>" code I went back and redid a bit of these patches.

Now "refresh <height>" works, presumably as it was originally intended to, and is independent of the m_refresh_from_block_height setting. I revived my old patch which required explicitly setting the refresh-height variable to use m_refresh_from_block_height. I've also added code for newly generated wallets to use the current blockchain height for m_refresh_from_block_height.

Also since "refresh <height>" now works, I've changed fast_refresh() to update the start_height. In that case, the short_chain_history is ignored by the /getblocks.bin RPC, so it's no longer necessary to recreate the short chain after fast_refresh finishes.

moneromooo-monero commented:
Oh I see, the short chain is used to determine the height if it wasn't passed. I hadn't known that :)

moneromooo-monero commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -664,6 +665,20 @@ bool simple_wallet::set_variable(const std::vector<std::string> &args)
+      }
+      else
+      {
+        uint32_t height = boost::lexical_cast<uint32_t>(args[1]);
```
uint64_t, other height params are 64 bit.

moneromooo-monero commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -1136,6 +1151,17 @@ bool simple_wallet::init(const boost::program_options::variables_map& vm)
+    if (!m_restore_height)
+    {
+      std::string heightstr = command_line::input_line("Restore from specific blockchain height (optional, default 0): ");
+      try {
```
if (std::cin.eof())
          return false;

moneromooo-monero commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -1136,6 +1151,17 @@ bool simple_wallet::init(const boost::program_options::variables_map& vm)
+          m_restore_height = boost::lexical_cast<uint64_t>(heightstr);
+      }
+      catch (boost::bad_lexical_cast &) {
+        fail_msg_writer() << tr("bad m_restore_height parameter:") << " " << heightstr;
```
Do you want to return false here ? Else, print a "defaulting to scanning from start" to avoid it looking like an actual unhandled error.

moneromooo-monero commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -3030,12 +3069,25 @@ int main(int argc, char* argv[])
+    tools::signal_handler::install([&wal, &quit](int) {
+      quit = true;
+      wal.stop();
+    });
```
That looks like a patch of mine that's already been merged. Merge error somewhere ?

hyc replied:
Probably. I'll grab current master and put it all together again.

hyc commented:
OK, cleaned up again as discussed on IRC. "refresh &lt;height&gt;" fetches hashes up to the specified height. All pulls after the first use the short_chain_history; start_height is reset to zero so it's ignored.

moneromooo-monero commented:
reviewed
